### PR TITLE
feat: switch to Aleph BFT (rebase)

### PR DESCRIPTION
**Pull request by joschisan**
This pr realises the switch from HBBFT to Aleph BFT which has been tracked in https://github.com/fedimint/fedimint/issues/2638

We note a slight regression here that aleph bft only supports federations of at least four guardians.

### CODE REVIEW

File: fedimint-server/Cargo.toml [hunk 1]
```diff
@@ -44,6 +45,11 @@ tokio-stream = "0.1.11"
 tokio-rustls = "0.23.4"
 tokio-util = { version = "0.7.4", features = [ "codec" ] }
 tracing-subscriber = { version = "0.3.16", features = [ "env-filter" ] }
+aleph-bft = { version = "0.20.5", default-features = false }
+bitcoin_30 = { package = "bitcoin", version = "0.30.0" }
```
Comment (dpc) [inline fedimint-server/Cargo.toml:49]: Do we really really need it? :D

File: recoverytool/src/main.rs [hunk 2]
```diff
@@ -187,9 +182,13 @@ async fn main() -> anyhow::Result<()> {
             ]);
 
             let db = Database::new(RocksDb::open(db).expect("Error opening DB"), decoders);
-            let mut dbtx = db.begin_transaction().await;
+            let _dbtx = db.begin_transaction().await;
 
-            let mut change_tweak_idx: u64 = 0;
+            let _change_tweak_idx: u64 = 0;
+
+            unimplemented!();
```
Comment (dpc) [inline recoverytool/src/main.rs:189]: ? Is there a TODO for it?

Comment (dpc) [reply recoverytool/src/main.rs:189]: We can no longer recover funds from a federation until this is impl. or do I misunderstand?

File: modules/fedimint-mint-client/src/backup.rs [hunk 3]
```diff
@@ -88,7 +88,7 @@ impl MintClientModule {
             notes,
             pending_notes,
             next_note_idx,
-            epoch_count,
+            epoch_count: fedimint_block_count,
```
Comment (dpc) [inline modules/fedimint-mint-client/src/backup.rs:91]: Nit: rename the `epoch_count` as well? (can be a follow-up)

File: fedimint-testing/src/db.rs [hunk 4]
```diff
@@ -164,7 +164,7 @@ fn open_temp_db_and_copy(
 }
 
 /// Helper function that recursively copies all of the contents from
-/// `src` to `dst`.
+/// `atomic_broadcast` to `dst`.
```
Comment (dpc) [inline fedimint-testing/src/db.rs:167]: Automatic rename went wrong, I guess.

File: fedimint-server/src/db.rs [hunk 5]
```diff
@@ -48,27 +72,33 @@ impl_db_lookup!(
...
+    db_prefix = DbKeyPrefix::SignedBlock,
+    notify_on_modify = true,
 );
-impl_db_lookup!(key = EpochHistoryKey, query_prefix = EpochHistoryKeyPrefix);
+impl_db_lookup!(key = SignedBlockKey, query_prefix = SignedBlockPrefix);
 
-#[derive(Debug, Encodable, Decodable, Serialize)]
-pub struct LastEpochKey;
+#[derive(Debug, Encodable, Decodable)]
+pub struct AlephUnitsKey(pub u64, pub u64);
```
Comment (dpc) [inline fedimint-server/src/db.rs:90]: I have no idea what it is, and it is composed of two unnamed integers. Ideally it needs named fields and docs.

File: fedimint-server/src/db.rs [hunk 6]
```diff
@@ -31,6 +32,29 @@ impl std::fmt::Display for DbKeyPrefix {
...
+pub struct SessionIndexKey;
+
+impl_db_record!(
+    key = SessionIndexKey,
+    value = u64,
+    db_prefix = DbKeyPrefix::SessionIndex,
+);
+
+#[derive(Clone, Debug, Eq, PartialEq, Encodable, Decodable, Serialize, Ord, PartialOrd)]
+pub struct AcceptedIndex(pub u64);
```
Comment (dpc) [inline fedimint-server/src/db.rs:45]: Would love to see docs here as well. I can figure out the code if I know what data means, but if I don't know what data means, I don't even know my own name.

File: fedimint-server/src/consensus/server.rs [hunk 7]
```diff
@@ -1,106 +1,51 @@
...
 
 /// Runs the main server consensus loop
 pub struct ConsensusServer {
     /// Delegate for processing consensus information
-    consensus: FedimintConsensus,
-    /// Receives event notifications from the `ConsensusApi` (triggers epochs)
-    api_receiver: Peekable<ReceiverStream<ApiEvent>>,
-    /// P2P connections for running consensus
-    connections: PeerConnections<EpochMessage>,
+    pub consensus: FedimintConsensus,
```
Comment (dpc) [inline fedimint-server/src/consensus/server.rs:38]: Donno how I feel about making everything `pub` here, but I guess it's OK.

File: fedimint-server/src/consensus/mod.rs [hunk 8]
```diff
@@ -107,105 +58,85 @@ impl FedimintConsensus {
...
+        peer_id: PeerId,
+    ) -> anyhow::Result<()> {
+        let _timing /* logs on drop */ = timing::TimeReporter::new("process_consensus_item");
 
-        dbtx.commit_tx_result()
+        let mut reader = std::io::Cursor::new(item);
+        let consensus_item = ConsensusItem::consensus_decode(&mut reader, &self.decoders())?;
+
+        let item_debug = debug::item_message(&consensus_item);
+        debug!("\n  Peer {peer_id}: {item_debug}");
```
Comment (dpc) [inline fedimint-server/src/consensus/mod.rs:101]: `\n`?

File: fedimint-server/src/atomic_broadcast/session.rs [hunk 9]
```diff
@@ -30,16 +31,20 @@ pub async fn run(
...
+    incoming_message_receiver: Receiver<Message>,
     outgoing_message_sender: Sender<(Message, Recipient)>,
     ordered_item_sender: mpsc::Sender<Option<(OrderedItem, oneshot::Sender<Decision>)>>,
-    signed_block_receiver: Receiver<SignedBlock>,
+    federation_api: WsFederationApi,
 ) -> anyhow::Result<SignedBlock> {
+    // if all nodes are correct the session will take 45 to 60 seconds. The
+    // more nodes go offline the longer the session will take to complete.
+    const EXPECTED_ROUNDS_PER_SESSION: usize = 45 * 4;
+    // this constant needs to be 3000 or less to guarantee that the session
```
Comment (dpc) [inline fedimint-server/src/atomic_broadcast/session.rs:42]: https://crates.io/crates/static_assertions maybe?

File: fedimint-server/Cargo.toml [hunk 10]
```diff
@@ -15,6 +15,7 @@ path = "src/lib.rs"
 [dependencies]
 fedimint-aead = { path = "../crypto/aead" }
 anyhow = "1.0.66"
+async-channel = "1.8.0"
```
Comment (dpc) [inline fedimint-server/Cargo.toml:18]: https://docs.rs/tokio/latest/tokio/sync/mpsc/index.html instead? Or you need mpmc? Or perf?

File: fedimint-core/src/block.rs [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        if let Some(root) = bitcoin30::merkle_tree::calculate_root(leaf_hashes) {
+            header[8..].copy_from_slice(&root.to_byte_array());
+        }
+
+        header
+    }
+}
+
+#[derive(Clone, Debug, Encodable, Decodable, Encode, Decode, PartialEq, Eq, Hash)]
+pub struct SchnorrSignature(pub [u8; 64]);
```
Comment (dpc) [inline fedimint-core/src/block.rs:53]: Should we serialize as hex string? See`fedimint_core::encoding::as_hex`. Just a thought.

File: fedimint-core/Cargo.toml [hunk 12]
```diff
@@ -48,6 +49,7 @@ miniscript = { version = "9.0.2", features = [ "compiler", "serde" ] }
 secp256k1-zkp = { version = "0.7.0", features = [ "use-serde", "bitcoin_hashes", "global-context" ] }
 macro_rules_attribute = "0.1.3"
 bitvec = "1.0.1"
+parity-scale-codec = { version = "3.5.0", features = ["derive"] }
```
Comment (dpc) [inline fedimint-core/Cargo.toml:52]: Another encoding? Any reason we can't use our built-in encodeable/decodable?

**Review by dpc — COMMENTED** [review]
Left some notes, but looking forward to see it land soon.

**Review by dpc — APPROVED** [review]

File: fedimint-server/Cargo.toml (hunk 1 above)
Comment (joschisan) [reply fedimint-server/Cargo.toml:49]: We need bitcoin_30 for calculating merle roots of blocks

File: recoverytool/src/main.rs (hunk 2 above)
Comment (joschisan) [reply recoverytool/src/main.rs:189]: Correct, no TODO yet.

File: fedimint-server/src/db.rs (hunk 5 above)
Comment (joschisan) [reply fedimint-server/src/db.rs:90]: Its the backup for aleph bft to recover from a crash

File: fedimint-server/src/db.rs (hunk 6 above)
Comment (joschisan) [reply fedimint-server/src/db.rs:45]: Generally the broadcast code is not final yet. Expect some refactors and changes over the coming weeks. Will do docs when design is more stable.

File: fedimint-server/src/consensus/server.rs (hunk 7 above)
Comment (joschisan) [reply fedimint-server/src/consensus/server.rs:38]: Also not final yet

File: fedimint-server/src/consensus/mod.rs (hunk 8 above)
Comment (joschisan) [reply fedimint-server/src/consensus/mod.rs:101]: copy and pasted that from the previous code

File: fedimint-server/src/atomic_broadcast/session.rs (hunk 9 above)
Comment (joschisan) [reply fedimint-server/src/atomic_broadcast/session.rs:42]: good idea!

File: fedimint-server/Cargo.toml (hunk 10 above)
Comment (joschisan) [reply fedimint-server/Cargo.toml:18]: We need mpmc

File: fedimint-core/src/block.rs (hunk 11 above)
Comment (joschisan) [reply fedimint-core/src/block.rs:53]: Why would we?

File: fedimint-core/Cargo.toml (hunk 12 above)
Comment (joschisan) [reply fedimint-core/Cargo.toml:52]: unfortunately aleph bft uses it. At least we only use it just in time only for right before a struct is fed into aleph bft.

File: fedimint-server/src/db.rs (hunk 5 above)
Comment (dpc) [reply fedimint-server/src/db.rs:90]: Put it in the comment / field names. :)

File: fedimint-core/src/block.rs (hunk 11 above)
Comment (dpc) [reply fedimint-core/src/block.rs:53]: We only use `serde` via jsonrpc where byte arrays are serialized as `[0, 123, 200, 234, ...]` which is terrible to look at and takes quite a bit of space on the wire.

**Pull request merged**